R/Solver.R

```diff
@@ -9,161 +9,198 @@
+  if (FT_rank < RankAT) {
+    ## Use SVD of F' to find basis for its kernel
+    FT_svd <- svd(
+      t(F_mat),
+      nu = 0,
+      nv = nrow(F_mat)
+    )
+    kerFT_basis <- FT_svd$v[, (FT_rank + 1):ncol(FT_svd$v), drop = FALSE]
+
+    ## Use QR decomposition on (A' basis | F' basis) to get Q,
+    ### whose first nullity(F') columns are an orthonormal basis for F',
+    ### the first nullity(A') columns of which are an orthonormal basis for A',
+    ### so the nullity(F')-nullity(A') columns between are an orthonormal basis for F'\A'
+    kerFT_only <- qr.Q(qr(cbind(kerAT_basis, kerFT_basis), LAPACK = FALSE))[, (dim(kerAT_basis)[2] + 1):(dim(kerFT_basis)[2]), drop = FALSE]
+    ### Re-normalize them to sum (in abs. value) to 1 and smooth near-zeros:
+    kerFT_norm <- apply(kerFT_only,
+                        2,
+                        function(col) ifelse(abs(col) < .Machine$double.eps, 0, col / sum(abs(col)))
+    )
+
+    ## Get observation weights of these F'\A' basis vectors
+    ATw.weights <- t(A_mat) %*% kerFT_only
+    ### Re-normalize them to sum (in abs. value) to 1 and smooth near-zeros:
+    # ATw_norm <- apply(ATw.weights, MARGIN=2,
+    #                     FUN=function(col) ifelse(abs(col) < .Machine$double.eps, 0, col/sum(abs(col))))
+    ## Use normalized DID weights to get observation weights:
+    ATw_norm <- apply(t(A_mat) %*% kerFT_norm,
+                      2,
+                      function(col) ifelse(abs(col) < .Machine$double.eps, 0, col)
+    )
+
+    ## Append F'\A' basis vector weights to Obs.weights and DID.weights as "Add.Obs.weights"
+    Obs.weights <- cbind(Obs.weights, Add.Obs.weights = ATw_norm)
+    DID.weights <- cbind(DID.weights, Add.Obs.weights = kerFT_norm)
+  } else if (FT_rank > RankAT) {
+    stop("F has greater rank than A. Please check inputs for accuracy.")
   }
+
+  if (DID_full) { ## Add DID-estimator-only weights (that do not affect observation weights) if requested
+    print("Note: the weights labelled Add.DID.weights affect the estimator weights but not the observation weights.")
+    DID.weights <- cbind(DID.weights, Add.DID.weights = kerAT_norm)
+  } else { ## Otherwise print a note
+    print("Note: returning only a single solution for the DID estimator weights, althoughs others may exist that yield equivalent observation weights.")
+  }
+
+  return(
+    list(
+      DID.weights = DID.weights,
+      Obs.weights = Obs.weights
+    ))
 }
 
 ## Helper function to create v vectors
 ### Input the total length of the vector (Length); i.e. the total number of unique Thetas
 ### And the indices of the values that should be non-zero (NonZero)
 ### If value==NULL (the default), the non-zero values will be given equal weight adding to 1
 ### Otherwise, specify the weight for each, in the same order as NonZero
-create_V <- function(Length,NonZero,
-                     Values=NULL) {
-  v <- rep(0,Length)
+create_V <- function(
+    Length,
+    NonZero,
+    Values = NULL) {
```

Comment:
Single-indent function definition to clearly separate the function name from its arguments.